website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md

```diff
@@ -0,0 +1,160 @@
+Not all operations support the comment field.
+For instance, operations like `insert` or `listCollections` do not support it.
+
+Another challenge with the `comment` field is its handling in some MongoDB drivers, where it is typed as a string.
+This means that, in principle, any string can be passed.
+In our scenario, where we allow FerretDB to receive and parse
+tracing data through the `comment` field, the JSON must be correctly encoded by the driver and then decoded by FerretDB.
+
+A more robust solution would involve having a dedicated field for passing request-related context,
+and it would be ideal to establish a standard for such fields.
+For instance, it could be a BSON document with particular tracing-related fields.
+This would provide a more reliable method for passing context to the database.
+
+## Example application
+
+Since such a standard does not yet exist, let's explore [an example application](https://gist.github.com/rumyantseva/3c6ef7c7dfc3fbdea8f94a31f4a17885)
+that interacts with FerretDB and passes the tracing context using the `comment` field.
+
+For simplicity, most error handling is omitted in the example below.
+
+```go
+collection := client.Database("testdb").Collection("customers")
+
+user := bson.D{
+    {Key: "name", Value: "John Doe"},
+    {Key: "email", Value: "john.doe@example.com"},
+}
+
+insertCtx, insertSpan := tracer.Start(ctx, "InsertCustomer")
+_, _ = collection.InsertOne(insertCtx, user)
+insertSpan.End()
+
+findCtx, findSpan := tracer.Start(ctx, "FindCustomer")
+filter := bson.D{{Key: "name", Value: "John Doe"}}
+
+var result bson.D
+_ = collection.FindOne(findCtx, filter).Decode(&result)
+
+findSpan.End()
+```
+
+In this setup, we create two spans: one for inserting a document and another for finding it.
+
+If we run this application, we will see that the spans for the `InsertCustomer` and `FindCustomer` operations are created:
+
+[![Trace without propagation](/img/blog/ferretdb-otel/without-propagation.png)](/img/blog/ferretdb-otel/without-propagation.png)
+
+Let's add more details.
+If we pass the tracing context through the `comment` field, we will see that the spans are linked.
+Let's modify the `FindCustomer` part:
+
+```go
+findCtx, findSpan := tracer.Start(ctx, "FindCustomer")
+filter := bson.D{{Key: "name", Value: "John Doe"}}
+
+traceID := findSpan.SpanContext().TraceID().String()
+spanID := findSpan.SpanContext().SpanID().String()
+
+traceContext := struct {
+    TraceID string `json:"traceID"`
+    SpanID  string `json:"spanID"`
+}{
+    TraceID: traceID,
+    SpanID:  spanID,
+}
+
+comment, _ := json.Marshal(map[string]interface{}{
+    "ferretDB": traceContext,
+})
+
+var result bson.D
+_ = collection.FindOne(findCtx, filter, options.FindOne().SetComment(string(comment))).Decode(&result)
+
+findSpan.End()
+```
+
+Now, if we run the application, we will see that the spans are linked and the `FindCustomer` span is a child span created on the FerretDB side:
```

Comment:
Possibly I'm misunderstanding, isn't FindCustomer now a parent span? FYI, I see usage of wording _parent_ instead of _child_ in docs such as https://opentelemetry.io/docs/concepts/context-propagation/ so we may want do the same.

Reply:
I think it should be "child", but it would help to clarify child _of what_

Reply:
Referencing with a name would help :+1: